### PR TITLE
Avoid scheduler/parser manager deadlock by using non-blocking IO

### LINKED ISSUE #7935: scheduler gets stuck without a trace
**Apache Airflow version**:

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`):

**Environment**:

- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:
**What happened**:

The scheduler gets stuck without a trace or error. When this happens, the CPU usage of scheduler service is at 100%. No jobs get submitted and everything comes to a halt. Looks it goes into some kind of infinite loop. 
The only way I could make it run again is by manually restarting the scheduler service. But again, after running some tasks it gets stuck. I've tried with both Celery and Local executors but same issue occurs. I am using the -n 3 parameter while starting scheduler. 

Scheduler configs,
job_heartbeat_sec = 5
scheduler_heartbeat_sec = 5
executor = LocalExecutor
parallelism = 32

Please help. I would be happy to provide any other information needed

**What you expected to happen**:

**How to reproduce it**:

**Anything else we need to know**:

Moved here from https://issues.apache.org/jira/browse/AIRFLOW-401

### PR DESCRIPTION
**Pull request by ashb**
Closes #7935, #15037 (I hope!) 

Thanks to @uranusjr for making the draft PR that pointed me to this solution in the first place.

 There have been long standing issues where the scheduler would "stop 
 responding" that we haven't been able to track down.

 Someone was able to catch the scheduler in this state in 2.0.1 and inspect
 it with py-spy (thanks, MatthewRBruce!)

 The stack traces (slightly shortened) were:

 ``` Process 6: /usr/local/bin/python /usr/local/bin/airflow scheduler 
 Python v3.8.7 (/usr/local/bin/python3.8) Thread 0x7FF5C09C8740 (active):
 "MainThread"
    _send (multiprocessing/connection.py:368)
    _send_bytes (multiprocessing/connection.py:411)
    send (multiprocessing/connection.py:206)
    send_callback_to_execute (airflow/utils/dag_processing.py:283)
    _send_dag_callbacks_to_processor (airflow/jobs/scheduler_job.py:1795)
    _schedule_dag_run (airflow/jobs/scheduler_job.py:1762)

 Process 77: airflow scheduler -- DagFileProcessorManager Python v3.8.7
 (/usr/local/bin/python3.8) Thread 0x7FF5C09C8740 (active): "MainThread"
    _send (multiprocessing/connection.py:368)
    _send_bytes (multiprocessing/connection.py:405)
    send (multiprocessing/connection.py:206)
    _run_parsing_loop (airflow/utils/dag_processing.py:698)
    start (airflow/utils/dag_processing.py:596)
 ```

 What this shows is that both processes are stuck trying to send data to 
 each other, but neither can proceed as both buffers are full, but since 
 both are trying to send, neither side is going to read and make more space
 in the buffer. A classic deadlock!

 The fix for this is two fold:

 1) Enable non-blocking IO on the DagFileProcessorManager side.

    The only thing the Manager sends back up the pipe is (now, as of 2.0)
   the DagParsingStat object, and the scheduler will happily continue
   without receiving these, so in the case of a blocking error, it is
   simply better to ignore the error, continue the loop and try sending
   one again later.

 2) Reduce the size of DagParsingStat

    In the case of a large number of dag files we included the path for
   each and every one (in full) in _each_ parsing stat. Not only did the
   scheduler do nothing with this field, meaning it was larger than it
   needed to be, by making it such a large object, it increases the
   likely hood of hitting this send-buffer-full deadlock case!

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: tests/utils/test_dag_processing.py [hunk 1]
```diff
@@ -521,17 +521,97 @@ def test_dag_with_system_exit(self):
...
+    @conf_vars({('core', 'load_examples'): 'False'})
+    @pytest.mark.backend("mysql", "postgres")
+    def test_pipe_full_deadlock(self):
+        import threading
+
+        dag_filepath = TEST_DAG_FOLDER / "test_scheduler_dags.py"
+
+        child_pipe, parent_pipe = multiprocessing.Pipe()
+
+        import socket
```
Comment (kaxil) [inline tests/utils/test_dag_processing.py]: Should we move the imports at the top of the file or are they intentionally here?

Comment (ashb) [reply tests/utils/test_dag_processing.py]: Yes, lets move it.

File: airflow/utils/dag_processing.py [hunk 2]
```diff
@@ -696,11 +706,20 @@ def _run_parsing_loop(self):
...
 
             dag_parsing_stat = DagParsingStat(
-                self._file_paths,
+                len(self._file_paths),
                 max_runs_reached,
                 all_files_processed,
             )
-            self._signal_conn.send(dag_parsing_stat)
+            try:
+                self._signal_conn.send(dag_parsing_stat)
```
Comment (XD-DENG) [inline airflow/utils/dag_processing.py]: I'm not sure: does applying a timeout context here makes sense?
The reason is it may try to send and get stuck but not fail explicitly. So to make it fail after specific timeout time may help.

May be a dumb question.

File: airflow/utils/dag_processing.py [hunk 3]
```diff
@@ -696,11 +706,20 @@ def _run_parsing_loop(self):
...
+            except BlockingIOError:
```
Comment (XD-DENG) [inline airflow/utils/dag_processing.py:714]: Any possibility of other error type that we may want to ignore? Again, may be a dumb question.

File: airflow/utils/dag_processing.py (hunk 2 above)
Comment (ashb) [reply airflow/utils/dag_processing.py]: Since these are local sockets, with the change to Non-blocking mode, it will either succeed, or fail entirely, "instantly", so we don't need a timeout, no.

File: tests/utils/test_dag_processing.py (hunk 1 above)
Comment (ashb) [reply tests/utils/test_dag_processing.py]: https://github.com/apache/airflow/pull/15112/commits/c70f6460d0bacc6798e890c88735db4cfef67285

File: airflow/utils/dag_processing.py (hunk 3 above)
Comment (ashb) [reply airflow/utils/dag_processing.py:714]: I don't think so, no. Anything else is an "error" condition, and we want to die (and the scheduler will notice and restart the manager process)

File: airflow/utils/dag_processing.py (hunk 2 above)
Comment (uranusjr) [reply airflow/utils/dag_processing.py]: It wouldn't be a bad idea to add a timeout, but that's extremely rarely an issue for local sockets (things need to get pretty wrong for that to happen at all). This can always be revisited if someone ever complains.

Comment (ashb) [reply airflow/utils/dag_processing.py]: On Linux these are AF_UNIX sockets too, so even harder to get it to not get instant fail/success in non-blocking mode I think

File: tests/utils/test_dag_processing.py [hunk 4]
```diff
@@ -521,17 +523,93 @@ def test_dag_with_system_exit(self):
...
+        )
+
+        try:
+            thread.start()
+
+            # If this completes without hanging, then the test is good!
+            manager._run_parsing_loop()
+            exit_event.set()
+        finally:
+            import logging
```
Comment (kaxil) [inline tests/utils/test_dag_processing.py]: This can go at the top too

File: tests/utils/test_dag_processing.py [hunk 5]
```diff
@@ -521,17 +523,93 @@ def test_dag_with_system_exit(self):
...
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_SNDBUF, 1024)
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_RCVBUF, 1024)
+            sock.detach()
+
+        exit_event = threading.Event()
+
+        # To test this behaviour we need something that continually fills the
+        # parent pipe's bufffer (and keeps it full).
+        def keep_pipe_full(pipe, exit_event):
+            import logging
```
Comment (kaxil) [inline tests/utils/test_dag_processing.py]: and this too

File: tests/utils/test_dag_processing.py [hunk 6]
```diff
@@ -521,17 +523,93 @@ def test_dag_with_system_exit(self):
...
+            signal_conn=child_pipe,
+            pickle_dags=False,
+            async_mode=True,
+        )
+
+        try:
+            thread.start()
+
+            # If this completes without hanging, then the test is good!
+            manager._run_parsing_loop()
```
Comment (kaxil) [inline tests/utils/test_dag_processing.py:603]: Should we add a timeout to this -- so that when the tests fails it does not hang the CI,

instead we timeout and raise and exception where we can explain further

Comment (ashb) [reply tests/utils/test_dag_processing.py:603]: Oh in tests, yes we probably should.

File: tests/utils/test_dag_processing.py [hunk 7]
```diff
@@ -521,17 +523,93 @@ def test_dag_with_system_exit(self):
...
 
         # Three files in folder should be processed
-        assert len(result.file_paths) == 3
+        assert result.num_file_paths == 3
 
         with create_session() as session:
             assert session.query(DagModel).get(dag_id) is not None
 
+    @conf_vars({('core', 'load_examples'): 'False'})
+    @pytest.mark.backend("mysql", "postgres")
```
Comment (ashb) [inline tests/utils/test_dag_processing.py:540]:
```suggestion
    @pytest.mark.backend("mysql", "postgres")
    @pytest.mark.timeout(60)
```

File: tests/utils/test_dag_processing.py [hunk 8]
```diff
@@ -20,7 +20,9 @@
 import os
```
Comment (ashb) [inline tests/utils/test_dag_processing.py:21]:
```suggestion
import logging
import os
```

File: tests/utils/test_dag_processing.py [hunk 9]
```diff
@@ -521,17 +523,93 @@ def test_dag_with_system_exit(self):
...
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_RCVBUF, 1024)
+            sock.detach()
+
+        exit_event = threading.Event()
+
+        # To test this behaviour we need something that continually fills the
+        # parent pipe's bufffer (and keeps it full).
+        def keep_pipe_full(pipe, exit_event):
+            import logging
+
```
Comment (ashb) [inline tests/utils/test_dag_processing.py]:
```suggestion
```

File: tests/utils/test_dag_processing.py [hunk 10]
```diff
@@ -521,17 +523,93 @@ def test_dag_with_system_exit(self):
...
+        try:
+            thread.start()
+
+            # If this completes without hanging, then the test is good!
+            manager._run_parsing_loop()
+            exit_event.set()
+        finally:
+            import logging
+
```
Comment (ashb) [inline tests/utils/test_dag_processing.py]:
```suggestion
```

File: tests/utils/test_dag_processing.py (hunk 6 above)
Comment (ashb) [reply tests/utils/test_dag_processing.py:603]: https://github.com/apache/airflow/pull/15112/commits/5e7405f5574b36e8bc9cf8df0d0d788ba7238ad4

**Review by kaxil — APPROVED** [review]

File: tests/utils/test_dag_processing.py (hunk 5 above)
Comment (ashb) [reply tests/utils/test_dag_processing.py]: https://github.com/apache/airflow/pull/15112/commits/5e7405f5574b36e8bc9cf8df0d0d788ba7238ad4

File: tests/utils/test_dag_processing.py (hunk 4 above)
Comment (ashb) [reply tests/utils/test_dag_processing.py]: https://github.com/apache/airflow/pull/15112/commits/5e7405f5574b36e8bc9cf8df0d0d788ba7238ad4

**Review by kaxil — APPROVED** [review]

File: tests/utils/test_dag_processing.py [hunk 11]
```diff
@@ -17,10 +17,13 @@
 # under the License.
 
 import multiprocessing
+import logging
```
Comment (ashb) [inline tests/utils/test_dag_processing.py]:
```suggestion
import logging
import multiprocessing
```

File: tests/utils/test_dag_processing.py [hunk 12]
```diff
@@ -521,17 +524,90 @@ def test_dag_with_system_exit(self):
...
+        for fd in (parent_pipe.fileno(),):
+            sock = socket.socket(fileno=fd)
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_SNDBUF, 1024)
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_RCVBUF, 1024)
+            sock.detach()
+
+        exit_event = threading.Event()
+
+        # To test this behaviour we need something that continually fills the
+        # parent pipe's bufffer (and keeps it full).
```
Comment (XD-DENG) [inline tests/utils/test_dag_processing.py]:
```suggestion
        # parent pipe's buffer (and keeps it full).
```

File: airflow/utils/dag_processing.py [hunk 13]
```diff
@@ -141,7 +141,7 @@ def waitable_handle(self):
 class DagParsingStat(NamedTuple):
     """Information on processing progress"""
 
-    file_paths: List[str]
+    num_file_paths: int
```
Comment (XD-DENG) [inline airflow/utils/dag_processing.py]: let's further challenge this part: why do we need `num_file_paths` (or the earlier `file_paths`) at all?

Where we really need this `DagParsingStat` is `self._sync_metadata`, in which we don't use `num_file_paths` (or the earlier `file_paths`)

Comment (XD-DENG) [reply airflow/utils/dag_processing.py]: Checking `self._sync_metadata` also leads me to something else:

https://github.com/apache/airflow/blob/1bec3b21266764f367431ab5d9a5b75f52b9b6d2/airflow/utils/dag_processing.py#L323-L326 is doing exactly the same thing in `self._process_message()`, which has been invoked just before L323-326. 

A clean-up refactor to do? Or I misunderstood something?

Comment (ashb) [reply airflow/utils/dag_processing.py]: Fair point yes, this is only used for tests and could possibly be removed. Let me take a look.

Comment (ashb) [reply airflow/utils/dag_processing.py]: Yes, done, was only used in one test, and that was strictly speaking not testing the right thing anymore, just a likely side-effect.

```diff
-        assert result.num_file_paths == 3
+        assert sum(stat.run_count for stat in manager._file_stats.values()) == 3
```

See 4d8da57eb for full diff.

Good call @XD-DENG

Comment (XD-DENG) [reply airflow/utils/dag_processing.py]: > Checking `self._sync_metadata` also leads me to something else:
> 
> https://github.com/apache/airflow/blob/1bec3b21266764f367431ab5d9a5b75f52b9b6d2/airflow/utils/dag_processing.py#L323-L326
> 
> is doing exactly the same thing in `self._process_message()`, which has been invoked just before L323-326.
> A clean-up refactor to do? Or I misunderstood something?

Thanks @ashb . How about this question?

We can address it separately of course, but want to confirm if I overlooked anything regarding this.

Comment (ashb) [reply airflow/utils/dag_processing.py]: Yeah, that looks like a bug. PR comin.

Comment (kaxil) [reply airflow/utils/dag_processing.py]: lol yeah, good catch @XD-DENG `self. _sync_metadata` is redundant here

**Review by XD-DENG — APPROVED** [review]
